src/tools/non-promotion-reasons.md

```diff
@@ -111,16 +163,546 @@ one or more of the following:
+
+{:.good}
+```dart
+f(Base b) {
+  final j = b._i;
+  if (i != null) {
+    int i = j; // OK
+  }
+}
+```
+
+#### Note about abstract fields
+
+If the getter in question is [abstract][], you might be able to enable field
+promotion and avoid the error by replacing the abstract getter with an [abstract field][].
+
+For example, this code produces an error because the getter `_i` can't be promoted
+and assigned to a non-nullable type:
+
+{:.bad}
+```dart
+abstract class Base {
+  int? get _i; // CONTEXT
+}
+
+class Derived extends Base {
+  @override
+  final int? _i;
+  Derived(this._i);
+}
+
+f(Base b) {
+  if (b._i != null) {
+    int i = b._i; // ERROR
+  }
+}
+```
+
+Changing the getter to an abstract field enables type promotion, removing the error:
+
+{:.good}
+```dart
+abstract class Base {
+  abstract final int? _i;
+}
+
+class Derived extends Base {
+  @override
+  final int? _i;
+  Derived(this._i);
+}
+
+f(Base b) {
+  if (b._i != null) {
+    int i = b._i; // OK
+  }
+}
+```
+
+{{site.alert.note}}
+Abstract fields are syntactically identical to abstract getters.
+The only behavior difference is in field promotion. 
+{{site.alert.end}}
+
+### External fields can't be promoted {#external}
+
+**The cause:**
+You're trying to promote a field, but the field is marked `external`.
+
+External fields don't promote because they are essentially external getters;
+their implementation is code from outside of Dart,
+so there’s no guarantee for the compiler that an external field 
+will return the same value each time it’s called.
+
+**Example:**
+
+{:.bad}
+```dart
+class C {
+  external final int? _externalField; // CONTEXT
+  C(this._externalField);
+
+  f() {
+    if (_externalField != null) {
+      int i = _externalField; // ERROR
+    }
+  }
+}
+```
+
+**Message:**
+
+```nocode
+'externalField' refers to an external field so it couldn’t be promoted.
+```
+
+**Solution:**
+
+Assign the external field's value to a local variable:
+
+{:.good}
+```dart
+class C {
+  external final int? _externalField;
+  C(this._externalField);
+
+  f() {
+    final j = this._externalField;
+    if (j != null) {
+      int i = j; // OK
+    }
+  }
+}
+```
+
+### Conflict with getter elsewhere in library {#getter-name}
+
+**The cause:**
+You're trying to promote a field,
+but another class in the same library contains
+a concrete getter with the same name.
+
+**Example:**
+
+{:.bad}
+```dart
+import 'dart:math';
+
+class Example {
+  final int? _overridden;
+  Example(this._overridden);
+}
+
+class Override implements Example {
+  @override
+  int? get _overridden => Random().nextBool() ? 1 : null; // CONTEXT
+}
+
+f(Example x) {
+  if (x._overridden != null) {
+    int i = x._overridden; // ERROR
+  }
+}
+```
+
+**Message:**
+
+```nocode
+'overriden' couldn’t be promoted because there is a conflicting getter in class 'Override'
+```
+
+**Solution**:
+
+*// what is the solution here? would you be able to rename them if they ARE related, too?*
```

Comment:
@stereotype441

Reply:
Good point. The only way to fix the error through renaming is to rename either the getter or the field but not both. If you've given them the same name on purpose because they're related (e.g. because one of them overrides the other, as in the above example), then renaming one of them but not the other will break the relationship between them, and your code will no longer do what you intend (it might not even compile).

Maybe we need to rewrite this section a bit. As it stands we've got the section "Conflict with getter elsewhere in library" that the user will see first, with an example where the classes are related, and then we have the subsection "Note about unrelated classes", with an example where the classes aren't related.

Maybe it would be better to lead with the second example, where the classes aren't related, and the solution is to do a rename. Then follow up with a subsection "But what if the classes are related?", and that subsection breaks the news to the user that if the two conflicting things are related, they're probably going to have to do the trick with a local final variable (as you've done in most of the other examples in this file).

Does that seem reasonable? I'm spitballing here.